File: backend/routes/extras.js
import { Router } from "express";
import { resolveLocation } from "../services/geocode.js";

const router = Router();

/**
 * GET /api/extras?location=<anything>
 * Additional API integrations for a location (assessment section 2.2):
 * - Map data: OpenStreetMap embed URL (no key) + Google Maps link
 * - YouTube videos about the location (requires optional YOUTUBE_API_KEY env var;
 *   degrades gracefully to a search link when no key is configured)
 */
router.get("/", async (req, res, next) => {
  try {
    const place = await resolveLocation(req.query.location);
    const { lat, lon, name } = place;

    // Map data - works with zero configuration
    const bbox = [lon - 0.05, lat - 0.03, lon + 0.05, lat + 0.03].join("%2C");
    const map = {
      provider: "OpenStreetMap",
      embedUrl: `https://www.openstreetmap.org/export/embed.html?bbox=${bbox}&layer=mapnik&marker=${lat}%2C${lon}`,
      googleMapsUrl: `https://www.google.com/maps/search/?api=1&query=${lat}%2C${lon}`,
      latitude: lat,
      longitude: lon,
    };

    // YouTube - real API when a key is present, graceful fallback when not
    const shortName = name.split(",")[0];
    let youtube;
    const key = process.env.YOUTUBE_API_KEY;
    if (key) {
      const yt = await fetch(
        `https://www.googleapis.com/youtube/v3/search?part=snippet&type=video&maxResults=4&q=${encodeURIComponent(
          shortName + " travel guide"
        )}&key=${key}`
      );
      if (yt.ok) {
        const data = await yt.json();
        youtube = {
          mode: "api",
          videos: (data.items || []).map((v) => ({
            title: v.snippet.title,
            channel: v.snippet.channelTitle,
            thumbnail: v.snippet.thumbnails?.medium?.url,
            url: `https://www.youtube.com/watch?v=${v.id.videoId}`,
          })),
        };
      }
    }
    if (!youtube) {
      youtube = {
        mode: "search-link",
        note: "Set YOUTUBE_API_KEY in backend/.env to fetch videos via the YouTube Data API.",
        searchUrl: `https://www.youtube.com/results?search_query=${encodeURIComponent(
          shortName + " travel guide"
        )}`,
      };
    }

    res.json({ location: { name, latitude: lat, longitude: lon }, map, youtube });
  } catch (err) {
    next(err);
  }
});

export default router;

File: backend/routes/extras.py
"""GET /api/extras?location=... - map data + YouTube videos (assessment section 2.2)."""
import os
from urllib.parse import quote

import requests
from fastapi import APIRouter, Query
from services.geocode import resolve_location

router = APIRouter()


@router.get("/api/extras")
def extras(location: str | None = Query(None)):
    place = resolve_location(location)
    lat, lon, name = place["lat"], place["lon"], place["name"]

    bbox = "%2C".join(str(v) for v in (lon - 0.05, lat - 0.03, lon + 0.05, lat + 0.03))
    map_data = {
        "provider": "OpenStreetMap",
        "embedUrl": f"https://www.openstreetmap.org/export/embed.html?bbox={bbox}&layer=mapnik&marker={lat}%2C{lon}",
        "googleMapsUrl": f"https://www.google.com/maps/search/?api=1&query={lat}%2C{lon}",
        "latitude": lat, "longitude": lon,
    }

    short = name.split(",")[0]
    q = f"{short} travel guide"
    youtube = None
    key = os.environ.get("YOUTUBE_API_KEY")
    if key:
        try:
            res = requests.get("https://www.googleapis.com/youtube/v3/search",
                               params={"part": "snippet", "type": "video", "maxResults": 4, "q": q, "key": key},
                               timeout=10)
            if res.ok:
                youtube = {"mode": "api", "videos": [
                    {"title": v["snippet"]["title"], "channel": v["snippet"]["channelTitle"],
                     "thumbnail": v["snippet"].get("thumbnails", {}).get("medium", {}).get("url"),
                     "url": f"https://www.youtube.com/watch?v={v['id']['videoId']}"}
                    for v in res.json().get("items", [])]}
        except requests.RequestException:
            youtube = None
    if youtube is None:
        youtube = {"mode": "search-link",
                   "note": "Set YOUTUBE_API_KEY to fetch videos via the YouTube Data API.",
                   "searchUrl": f"https://www.youtube.com/results?search_query={quote(q)}"}

    return {"location": {"name": name, "latitude": lat, "longitude": lon}, "map": map_data, "youtube": youtube}
